src/coreclr/jit/optimizer.cpp

```diff
@@ -9316,6 +9317,188 @@ bool OptBoolsDsc::optOptimizeBoolsCondBlock()
     return true;
 }
 
+//-----------------------------------------------------------------------------
+//  optOptimizeCompareChainCondBlock:  Create AND chain when when both m_b1 and m_b2 are BBJ_COND.
+//
+//  Returns:
+//      true if chain optimization is done and m_b1 and m_b2 are folded into m_b1, else false.
+//
+//  Assumptions:
+//      m_b1 and m_b2 are set on entry.
+//
+//  Notes:
+//
+//      This aims to reduced the number of conditional jumps by joining cases when multiple
+//      conditions gate the execution of a block. For example:
+//          If ( a > b || c == d) { x = y; }
+//      Will become the following. Note that the second condition is inverted.
+//
+//      ------------ BB01 -> BB03 (cond), succs={BB02,BB03}
+//      *  JTRUE
+//      \--*  GE a,b
+//
+//      ------------ BB02 -> BB04 (cond), preds={BB01} succs={BB03,BB04}
+//      *  JTRUE
+//      \--*  NE c,d
+//
+//      ------------ BB03, preds={BB01, BB02} succs={BB04}
+//      *  ASG x,y
+//
+//      These operands will be combined into a single AND chain in the first block (with the first
+//      condition inverted).
+//
+//      ------------ BB01 -> BB03 (cond), succs={BB03,BB04}
+//      *  JTRUE
+//      \--*  AND
+//         +--*  LT a,b
+//         \--*  NE c,d
+//
+//      ------------ BB03, preds={BB01} succs={BB04}
+//      *  ASG x,y
+//
+//
+//      This will also work for statements with else cases:
+//          If ( a > b || c == d) { x = y; } else { x = z; }
+//      Here BB04 will contain the else ASG. Both BB04 and BB05 will unconditionally jump to BB05.
+//
+//      ------------ BB01 -> BB03 (cond), succs={BB03,BB04}
+//      *  JTRUE
+//      \--*  AND
+//         +--*  LT a,b
+//         \--*  NE c,d
+//
+//      ------------ BB03, preds={BB01} succs={BB05}
+//      *  ASG x,y
+//
+//      ------------ BB04, preds={BB01} succs={BB05}
+//      *  ASG x,z
+//
+//
+//      Multiple conditions can be chained together. This is due to the optimization reverse
+//      iterating through the blocks. For example:
+//          If ( a > b || c == d || e < f ) { x = y; }
+//      The first pass will combine "c == d" and "e < f" into a chain. The second pass will then
+//      combine the "a > b" with the earlier chain, giving:
+//
+//      ------------ BB01 -> BB03 (cond), succs={BB03,BB04}
+//      *  JTRUE
+//      \--*  AND
+//         +--*  AND
+//            +--*  NE c,d
+//            +--*  GE e,f
+//         \--*  LT a,b
+//
+//      ------------ BB03, preds={BB01} succs={BB04}
+//      *  ASG x,y
+//
+//
+//      Conditions connected by && are not yet checked for. For example:
+//          If ( a > b && c == d ) { x = y; }
```

Comment:
Really what this optimization does is change short circuiting into bitwise ops, is that correctly understand? I.e. this pass just changes `a && b` into `a & b` and `a || b` into `a | b` when `a` and `b` are relops? But for some reason the former case is not supported?

Reply:
>Do you expect to work on this in the near future?

Today I'm not sure if or how we would should do it.

The way block flows work is that we jump on the inverse of the condition.

In existing code:
`If ( a > b ) { x = y; }`
becomes:
```
BB1: if (a <= b) jump to BB3
BB2: x=y
BB3: ....
```


With this PR:
`If ( a > b || c == d || e >= f) { x = y; }`
This can become an AND chain by testing the false of all the conditions:
```
BB1: if (a <= b && c != d && e < f) jump to BB3
BB2: x=y;
BB3: ....
```
(GCC and LLVM do this).


Consider:
`If ( a > b && c == d && e >= f) { x = y; }`
The only way to effectively use an combine the conditions here is:
```
BB1: If ( a > b && c == d && e >= f) jump to BB3
BB2: ....; return
BB3: x=y; jump to BB2
```
or if the remainder of the function is small enough:
```
BB1: If ( a > b && c == d && e >= f) jump to BB3
BB2: ....; return
BB3: x=y;
BB4: Repeat everything in BB2
```
(GCC does this, whilst LLVM keeps with what ryujit does today, which is have lots of branches)


I can't see a situation where
`If ( a > b && c == d || e >= f) { x = y; }`
becomes a single chain.

Reply:
I might be misunderstanding then. My question is: Do we have a situation where the JIT might handle something on the form:
```csharp
if (a || b || c)
{
  A();
}
else
{
  B();
}
```

but not something of the form:
```csharp
if (!a && !b && !c)
{
  B();
}
else
{
  A();
}
```

(In this case `a`, `b`, `c` would be some relops)? That's the case I was thinking of, and it seems problematic if these two isomorphic programs are not handled equivalently by the JIT. For example, the JIT may reverse one into the other at various points of flowgraph transformations, I believe.

Reply:
It'd handle either ok, but then be unable to convert one into the other. Fixing the fgopt should sort that out.